### ISSUE TITLE
Access from external apps

### ISSUE THREAD

**Issue by moovida** [arpitkh96/AmazeFileManager#212]
I am trying to use Amaze as default file browser from within my app.
Amaze already listens to the ACTION_GET_CONTENT intent.

It would be amazing to be able to be able to:

* pass a title to the dialog that picks the file instead of having the generic Toast telling to pick a file. 
* having a file filter for viewing files by extension (i.e. mimetype)
* be able to pick a folder instead of a file

and maybe not so important, but definitely a cherry on the cake :-) :
 
* have Amaze show a custom icon for my app's own filetypes

I'm just putting this here supposing that this is the place where to place feature wishes.

**Comment by vishal0071** [comment]
What is the extension for your filetypes?

**Comment by moovida** [comment]
There are a few of them:

* gpx: a quite known gps data format
* sqlite: for sqlite and spatialite databases
* gpap: which is the extension of the geopaparazzi app, which would be nifty to be able to supply an icon for


While trying to create a small pull request at least for the simple and boring parts I would be able to do, I noticed that one more thing that would make this amazing. Evaluate the contents that the user sets in:

 intent.setDataAndType(uri, mimeType);

Where other file managers take the uri as the path from which to start browsing and the mimetype... well, I couldn't find any filemanager that properly evaluates the mimetype to a filter. :-) 
But it would still be great to filter on that.

**Comment by vishal0071** [comment]
Can you give the icon for these mimeTypes?

**Comment by moovida** [comment]
Yes, would be happy to. Would you prefer svg or png? If png, which resolution?

**Comment by vishal0071** [comment]
You can mail me the svgs of sqlite and geopaparazzi. I don't think putting copyrighted icons in app will be a good idea. What kind of license do they have?

**Comment by moovida** [comment]
Hi @Geyup . In my app I have the need, for some tasks, to select files from the device's filesystem.
Until now I had implemented a simple file browsing activity, which would allow me to:
* open the browser in a certain folder
* select a file or a folder
* show icons for certain filetypes
That sounds nice, but all in all it was just a very simple activity with all the limitations of the case.

So I have been looking for an file manager app that exposes such functionalities as intents. The well known ES file explorer has a part of these, but it is not open source, which is important for me, being my app also open.

I did some testing, but without a dedicated set on intents, it all isn't very userfriendly. So right now I will keep both options in the hope that Amaze at some point will support these functionalities.

**Comment by Geyup** [comment]
Let me rephrase the question... :)

What application are you talking about, and what is it you are doing? By doing, I don't mean the actions you are accomplishing, by what you described, I mean, why do you need to do things in this manner, because as an example, you are moving files from local to remote, or syncing etc...

You need this because you are working remotely, or locally?

Ok, hope this explains things better...

**Comment by moovida** [comment]
Sorry for the late reply @Geyup , and the misunderstanding :-)

To answer your question, I am working mainly with local files. Remote sync is a different piece.
For example I have an engineer going in the field studying a river. He prepares a couple of maps with different themes. Out in the field (offline), he wants to switch between visualizations and load maps. In that case it can choose from local files.

Not sure I have been more clear, than before, but I hope yes :-)

**Comment by Geyup** [comment]
The things you said before were not clear.

You keep saying your ``app`` and you're not telling us what this app is?

Browsers are for surfing the Internet, so there's confusion.

You mention select a file or folder, file mangers already do this.

Next you talk about you working locally, but an engineer out in the field, saying' ```In that case it can choose from local files``` So now I'm assuming communication between local and remote?

Sounds like you need an application with Internet connectivity?

**Comment by moovida** [comment]
:-( I am definitely not making myself clear here. I apologize to the project leads for spamming their issue tracker with this. 

Example:

* my app is a survey app
* for each survey I can load a project
* my project files are created locally on the device
* when I start a survey I open the project of that survey, which is on the device
* before it looked like this: http://geopaparazzi.github.io/geopaparazzi/main_view_menu.html#load-an-existing-project
* same as for project files, it could be other files to import/select from the filesystem
* all this is needed for offline, device-local use

**Comment by Geyup** [comment]
I'm deleting all my posts and I suggest you do the same, and let's try this one more time! LOL

Explain it to everyone like they were 12 years old, what it is you'd like Amaze to do for you?

A very child like explaination is what's really needed here... :)

**Comment by moovida** [comment]
Think of any java desktop app that needs to access files on the filesystem. The jre supplies file choosers and folder choosers. 
It is as simple as that: I need file and folder choosers. Things I can launch supplying a startfolder, an extention to filter, a flag to define if I want to choose a file or folder, a title to give to the filesystembrowser that will be opened for the user.

Maybe that was the missleading word. When I was writing browser I meant a filesystembrowser, as for example thunar or dolphin on linux.

**Comment by Geyup** [comment]
Actually any java desktop app that needs access to filesystems is to broad a scope of explaination, it's tells nothing specific.

All along I've been trying to get you to be ```Specific```, and if you don't know how to be specific with the technology, explaining in understanable technology terms, then a simple example of what you want.

Thunar and Dolphin are flie managers, that is specific.

Amaze is a file manager, just like Thunar, Dolphin, of course each one might have slightly different features. Dolphin has more because it's a more ```bloated``` file manger.

Depending on the author/developer(s) intentions it can range in scope, I'm sure you understand this.

The reason why I've been asking so much is to figure out where you are coming from.

You do undertand that Amaze's goals seem to be a ```Simple``` & ```Lightweight``` filemanger?

So it's going to depend on the developer, if that fits within that goals.

I don't mean to be rude, maybe the developer has more of an understanding, but I'm a Linux geek and I even can't make heads or tails out of your words. You either need to understand the technology, or explain it in just simple ideas what you want to accomplish.

As example, a ```file chooser``` or ```folder chooser```, there's no such woding in comptuer technolgy.
With a file manger you can choose already file and folders, not sure what you mean by ```choosing```?

Amaze is very simple so there's not much to do, other then create, rename or delete, or move and copy, from what I'm recalling from memory.

I get that you'd like some form of a database..., sounds like the ability to also search by extension.

Pick a folder instead of a file makes no sense, I don't undertand this...

Anyhow, I've tried to be helpful, to at least try and point you in the direction of something else. If you haven't found anything on the Android system that will do what you want, you probably need to move over to Apple, Window or Linux.

Guess at what I'm seeing here, you'd be lucky if this was completed in 6 months...

**Comment by moovida** [comment]
I am not sure about why we are not understanding each other, but to use a common file explorer for android, here is what I would like to have:
http://www.estrongs.com/res/develop_en.htm

And Amaze already exposes an intent to select files. It just doesn't have folder and filter by extention. That is all. In this issue I was doing nothing else than opening a feature request. The author is free to ignore it or implement it if of interest. 

> Anyhow, I've tried to be helpful, to at least try and point you in the direction of something else. If you 
> haven't found anything on the Android system that will do what you want, you probably need to move 
> over to Apple, Windows or Linux.

I already found something that does this but is not open source and I prefer open source.

**Comment by Geyup** [comment]
Those screen shots are just basic File Management...

What you're looking for is a more ```Feature Rich``` file manager, Amaze seems to have simplicty in mind, just doing absolute basics...

ES isn't giving you enough of what you want, or no other file managers in Android?

**Comment by moovida** [comment]
Amaze is exactly what I want and what I use as file manager. I don't need it to be more feature rich.

I just would like for it to expose intents for ```basic File Management``` to use it from within other apps.

**Comment by Geyup** [comment]
Using from within other apps would depend on the developers idea of light weight, basic or not...

**Comment by davidbilly** [comment]
Hi @vishal0071 , is it the amaze fm able do folder picker? so far i just saw file picker. thanks for the awesome open source 👍

**Comment by vishal0071** [comment]
No it cant @davidbilly but I can try. Create a new issue about this.

**Comment by davidbilly** [comment]
@vishal0071 Thanks for the consider :D  i will create new issue. thank you very much!
